src/uct/ib/base/ib_device.c

```diff
@@ -549,12 +552,135 @@ ucs_status_t uct_ib_device_query(uct_ib_device_t *dev,
     return UCS_OK;
 }
 
+#if defined(__x86_64__)
+static int uct_ib_device_cleanup_proc(void* arg) {
+    uct_ib_device_nb_close_ctx *ctx = arg;
+    static const char *process_name = "ucx_cleanup";
+    struct rlimit nofile;
+    int ret = -EINTR;
+    int fd;
+    char c;
+
+    prctl(PR_SET_NAME, (uintptr_t)&process_name, 0, 0, 0);
+
+    getrlimit(RLIMIT_NOFILE, &nofile);
+    for (fd = 0; fd < nofile.rlim_cur && !ctx->quit; fd++) {
+        if ((fd != ctx->cmd_fd) && (fd != ctx->pipefds[0])) {
+            ucs_x86_syscall(SYS_close, fd);
+        }
+    }
+
+    /* loop until pipe closed - either parent terminated or closing device */
+    while (!ctx->quit && ret == -EINTR) {
+        ret = ucs_x86_syscall(SYS_read, ctx->pipefds[0], &c, 1);
+    }
+
+    return 0;
+}
+#endif
+
+static ucs_status_t uct_ib_device_init_nb_close_ctx(uct_ib_device_t *dev)
+{
+#if defined(__x86_64__)
+    const int stack_size = 32 * 1024;
+    const int tls_size = 1024;
+    uct_ib_device_nb_close_ctx *ctx;
+    ucs_status_t status;
+    int ret;
+
+    ctx = ucs_calloc(1, sizeof(*ctx), "ibv cleanup ctx");
+    if (ctx == NULL) {
+        ucs_error("cleanup context allocation failure");
+        status = UCS_ERR_NO_MEMORY;
+        goto err;
+    }
+
+    ctx->stack_size = stack_size;
+    status = ucs_mmap_alloc(&ctx->stack_size, &ctx->stack, 0
+                            UCS_MEMTRACK_NAME("ibv cleanup stack"));
+    if (status != UCS_OK) {
+        ucs_error("cleanup stack allocation failure");
+        goto err_alloc;
+    }
+
+    ctx->tls_size = tls_size;
+    status = ucs_mmap_alloc(&ctx->tls_size, &ctx->tls, 0
+                            UCS_MEMTRACK_NAME("ibv cleanup tls"));
+    if (status != UCS_OK) {
+        ucs_error("cleanup tls allocation failure");
+        goto err_alloc_tls;
+    }
+
+    ret = pipe(ctx->pipefds);
+    if (ret) {
+        ucs_error("cleanup pipe allocation failure");
+        status = UCS_ERR_IO_ERROR;
+        goto err_pipe;
+    }
+
+    ctx->cmd_fd = dev->ibv_context->cmd_fd;
```

Comment:
initialize all simple ctx fields in the beginning, then do all memory allocations
ctx->tls_size =...
ctx->stack_size = ...
ctx->cmd_fd = ..
ctx->pid = ..